src/main/java/seedu/address/logic/commands/UpdateCommand.java

```diff
@@ -71,78 +72,105 @@ public CommandResult execute(Model model, CommandHistory history) throws Command
+            throw new CommandException(MESSAGE_MISSING_EXPIRY);
         }
-        return updatedBatch;
     }
 
     /**
-     * Creates and returns a {@code Medicine} with the details of {@code medicineToUpdate}
-     * updated with {@code toUpdate}.
+     * Returns a new {@code Batch} with existing batch's {@code Expiry} added to {@code newBatchDetails}.
+     */
+    private Batch getExistingBatchExpiry(Batch batchToUpdate) {
+        return new Batch(newBatchDetails.getBatchNumber(), batchToUpdate.getExpiry(), newBatchDetails.getQuantity());
+    }
+
+    /**
+     * Returns a {@code Medicine} with the details of {@code medicineToUpdate} updated with {@code updatedBatch}.
      */
-    Medicine createUpdatedMedicine(Medicine medicineToUpdate, Batch updatedBatch) {
-        Map<BatchNumber, Batch> updatedBatches = getNewBatches(medicineToUpdate, updatedBatch);
-        Quantity updatedQuantity = getNewQuantity(medicineToUpdate, updatedBatch);
-        Expiry updatedExpiry = getNewExpiry(updatedBatches);
+    Medicine getUpdatedMedicine(Medicine medicineToUpdate, Batch batchToUpdate, Batch updatedBatch) throws
+            CommandException {
+        Map<BatchNumber, Batch> updatedBatches = getNewMedicineBatches(medicineToUpdate, updatedBatch);
+        Quantity updatedQuantity = getNewMedicineQuantity(medicineToUpdate, batchToUpdate, updatedBatch);
+        Expiry updatedExpiry = getNewMedicineExpiry(medicineToUpdate, updatedBatch, updatedBatches);
 
         return new Medicine(medicineToUpdate.getName(), updatedQuantity, updatedExpiry, medicineToUpdate.getCompany(),
                 medicineToUpdate.getTags(), updatedBatches);
     }
 
-    Map<BatchNumber, Batch> getNewBatches(Medicine medicineToUpdate, Batch updatedBatch) {
+    Map<BatchNumber, Batch> getNewMedicineBatches(Medicine medicineToUpdate, Batch updatedBatch) {
         HashMap<BatchNumber, Batch> newBatches = new HashMap<>(medicineToUpdate.getBatches());
-        if (updatedBatch.getQuantity().value.equals("0")) {
-            newBatches.remove(updatedBatch.getBatchNumber());
-        } else {
+        if (updatedBatch.hasNonZeroQuantity()) {
             newBatches.put(updatedBatch.getBatchNumber(), updatedBatch);
+        } else {
+            newBatches.remove(updatedBatch.getBatchNumber());
         }
         return newBatches;
     }
 
-    Quantity getNewQuantity(Medicine medicineToUpdate, Batch updatedBatch) {
+    Quantity getNewMedicineQuantity(Medicine medicineToUpdate, Batch batchToUpdate, Batch updatedBatch)
+            throws CommandException {
         int quantity = medicineToUpdate.getQuantity().getNumericValue();
-        Batch batchToUpdate = medicineToUpdate.getBatches().get(updatedBatch.getBatchNumber());
 
         if (batchToUpdate != null) {
             quantity -= batchToUpdate.getQuantity().getNumericValue();
         }
         quantity += updatedBatch.getQuantity().getNumericValue();
+
+        if (quantity > 999999999) {
```

Comment:
Use Java's `Integer.MAX_VALUE` if the intention is to prevent integer overflow.

Reply:
You're right! Thanks for pointing it out. I will amend it by next PR.